Stats: Now only valid probabilities can be entered in Bernoulli, Coin and FiniteRV

supreet11agrawal opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #15903 

#### Brief description of what is fixed or changed
frv_types earlier allowed non-valid probability distribution in case of Bernoulli, Coin and FiniteRV methods. After the modification, only probabilites with value in between 0 to 1 can be entered. Also in FiniteRV, earlier a random variable with non valid probability density was allowed. New changes prohibit such shortcomings.

Example:

Bernoulli
```
>>> from sympy.stats import Bernoulli, density
>>> from sympy import S
>>> X = Bernoulli('X', S(5)/4) #probability can't be >1 but 5/4 here
```
 Earlier this worked

Now it gives
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy/stats/frv_types.py", line 200, in Bernoulli
    return rv(name, BernoulliDistribution, p, succ, fail)
  File "sympy/stats/frv_types.py", line 28, in rv
    density = cls(*args)
  File "sympy/stats/frv_types.py", line 172, in __new__
    raise ValueError("'p' must be: 0 <= p <= 1 . p = %s" % str(p))
ValueError: 'p' must be: 0 <= p <= 1 . p = 5/4
```
Similarily, Coin and FiniteRV modified. Can be tested using invalid probability distribution

For FiniteRV, density's each element should be between 0 and 1 and also total density should be 1. This is now implemented.



#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
*  stats
    * fixed bugs in finite discrete random variable types 
<!-- END RELEASE NOTES -->

supreet11agrawal committed:
Stats: Now only valid probabilities can be entered in Bernoulli, Coin and FiniteRV

frv_types earlier allowed non-valid probability distribution in case of Bernoulli,
Coin and FiniteRV methods. After the modification, only probabilites with value
in between 0 to 1 can be entered. Also in FiniteRV, earlier a random variable
with non valid probability density was allowed. New changes prohibit such shortcomings.

Example:

Bernoulli

>>> from sympy.stats import Bernoulli, density
>>> from sympy import S
>>> X = Bernoulli('X', S(5)/4) #probability can't be >1 but 5/4 here
 Earlier this worked

Now it gives

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy/stats/frv_types.py", line 200, in Bernoulli
    return rv(name, BernoulliDistribution, p, succ, fail)
  File "sympy/stats/frv_types.py", line 28, in rv
    density = cls(*args)
  File "sympy/stats/frv_types.py", line 172, in __new__
    raise ValueError("'p' must be: 0 <= p <= 1 . p = %s" % str(p))
ValueError: 'p' must be: 0 <= p <= 1 . p = 5/4

Similarily, Coin and FiniteRV modified
Can be tested using invalid probability distribution

For FiniteRV, density's each element should be between 0 and 1 and also
total density should be 1. This is now implemented.

supreet11agrawal committed:
Merge branch 'master' into 15903_probability

supreet11agrawal commented:
@Upabjojr  @jksuom Can you please review?

supreet11agrawal committed:
Update frv_types.py

jksuom commented:
There are some repeated imports, otherwise this looks good. There could also be a test for the issue.

supreet11agrawal commented:
Which imports are you referring to? I don't see any repeated ones
I will add tests as well
Thanks +1:

jksuom commented:
This is from frv_types.py:
```
from sympy import (S, sympify, Rational, binomial, cacheit, Integer,
        Dict, Basic, KroneckerDelta, Dummy)
from sympy.concrete.summations import Sum
from sympy.core.compatibility import as_int, range
from sympy.core.logic import fuzzy_not, fuzzy_and
from sympy.stats.frv import (SingleFinitePSpace, SingleFiniteDistribution)
from sympy.concrete.summations import Sum
from sympy import (S, sympify, Rational, binomial, cacheit, Integer,
        Dict, Basic, KroneckerDelta, Dummy)
```

supreet11agrawal committed:
Update frv_types.py

supreet11agrawal commented:
Ohkk! Sorry, don't know how I missed that

supreet11agrawal commented:
@jksuom Is that okay?
I will add tests in a different PR

supreet11agrawal commented:
Mostly by tomorrow

smichr commented on sympy/stats/frv_types.py:
```diff
@@ -167,6 +174,15 @@ def Die(name, sides=6):
+        p_sym = sympify(p)
+
+        if fuzzy_not(fuzzy_and((p_sym.is_nonnegative, (p_sym - 1).is_nonpositive))):
+            raise ValueError("'p' must be: 0 <= p <= 1 . p = %s" % str(p))
```
or "p = %s is not in range [0, 1]."

smichr commented on sympy/stats/frv_types.py:
```diff
@@ -36,6 +36,13 @@ def dict(self):
+                raise ValueError("Probability at a point must be between 0 and 1")
+        sum_sym = sum(density.values())
+        if sum_sym != 1:
+            raise ValueError("Total Probability must be equal to 1")
```
add a period after 1

supreet11agrawal committed:
Update frv_types.py

supreet11agrawal commented:
I have done the necessary changes. Kindly merge this PR

oscarbenjamin commented:
The tests should be added before the PR is merged.

supreet11agrawal commented:
Ok.Will do that

supreet11agrawal committed:
Added tests in frv_types

Added tests to check if entered probability is valid or not.
Methods changed:- Bernoulli and FiniteRV
As coin uses Bernoulli Distribution, hence tests for coin are added as well

supreet11agrawal commented:
@jksuom I have added the tests. However, as I am new to Sympy I am not sure if I have added them correctly or if I should add more. Tell me if a change is required

jksuom commented:
Thanks, this looks good to me. The last test line is a bit longish but that can be fixed later.
